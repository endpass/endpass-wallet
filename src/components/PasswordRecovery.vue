<template>
  <div>
    <v-form
      :is-form-valid="isFormValid"
      @submit="toggleConfirmModal"
      data-test="password-recovery-form"
    >
      <label class="label">Password Recovery</label>
      <v-input
        v-validate="'required|seed_phrase'"
        v-model="seedPhrase"
        :error="errors.first('seedPhrase')"
        :disabled="isLoading"
        name="seedPhrase"
        placeholder="Seed phrase"
        data-test="input-seed-phrase"
      />
      <v-password
        v-validate="'required|min:8'"
        v-model="password"
        :error="errors.first('password')"
        :disabled="isLoading"
        name="password"
        placeholder="New password"
        data-test="input-password"
      />
      <v-button
        ref="button"
        :loading="isLoading"
        :disabled="!isFormValid"
        class-name="is-primary is-medium"
        data-test="recover-button"
      >
        Recover
      </v-button>
    </v-form>

    <confirm-modal
      v-if="isConfirmModal"
      @confirm="handleConfirm"
      @close="toggleConfirmModal"
    >
      <p class="subtitle">
        All keystore wallets will be deleted. They can be restored from a
        private key or seed phrase on the import page.
      </p>
    </confirm-modal>
  </div>
</template>

<script>
import { mapActions } from 'vuex';

import ConfirmModal from '@/components/modal/ConfirmModal';
import formMixin from '@/mixins/form';
import modalMixin from '@/mixins/modal';

export default {
  name: 'PasswordRecovery',
  data: () => ({
    seedPhrase: null,
    password: null,
    isLoading: false,
  }),
  methods: {
    ...mapActions('accounts', ['recoverWalletsPassword']),
    async handleConfirm() {
      try {
        this.isLoading = true;
        this.toggleConfirmModal();

        await this.recoverWalletsPassword({
          seedPhrase: this.seedPhrase,
          password: this.password,
        });
      } catch (error) {
        this.$notify({
          title: 'Error recovering wallet password',
          text:
            'An error occurred while recovering wallet password. Please try again.',
          type: 'is-danger',
        });
      } finally {
        this.seedPhrase = null;
        this.password = null;
        this.isLoading = false;
      }
    },
  },
  mixins: [formMixin, modalMixin],
  components: {
    ConfirmModal,
  },
};
</script>
