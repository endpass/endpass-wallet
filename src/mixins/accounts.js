// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
