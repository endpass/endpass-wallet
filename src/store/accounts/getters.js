import get from 'lodash/get';
import { fromWei, hexToBytes } from 'web3-utils';
import { BigNumber } from 'bignumber.js';
import keystoreHDKeyVerify from '@endpass/utils/keystoreHDKeyVerify';
import keystoreHDWallet from '@endpass/utils/keystoreHDWallet';
import isV3 from '@endpass/utils/isV3';

const wallet = state => get(state.wallets, state.address);

const accountAddresses = state =>
  Object.keys(state.wallets).map(item => item.toLowerCase());

const addressBuffer = state =>
  state.address && hexToBytes(state.address.toLowerCase());

const isPublicAccount = (state, getters) => !!get(getters.wallet, 'isPublic');

const isHardwareAccount = (state, getters) =>
  !!get(getters.wallet, 'isHardware');

const balance = (state, getters, rootState, rootGetters) => {
  if (!state.balance) return null;

  const pendingBalance = rootGetters['transactions/pendingBalance'];
  const balanceWei = BigNumber(state.balance || '0')
    .minus(pendingBalance)
    .toFixed();

  return fromWei(balanceWei);
};

const isHDv3WalletByType = state => walletType => {
  const cache = state.hdCacheByType[walletType];
  if (!cache) return false;
  const { xpub, v3KeyStore } = cache;
  const isPublic = keystoreHDKeyVerify.isExtendedPublicKey(xpub);
  const isV3Keystore = isV3(v3KeyStore);

  return isPublic && isV3Keystore;
};

const cachedXpubByType = state => walletType => {
  const cache = state.hdCacheByType[walletType];
  return get(cache, 'xpub');
};

const cachedHdV3KeyStoreByType = state => walletType => {
  const cache = state.hdCacheByType[walletType];
  return get(cache, 'v3KeyStore');
};

const getHdWalletBySeed = () => seedPhrase =>
  keystoreHDWallet.createHDWalletBySeed(seedPhrase);

export default {
  wallet,
  addressBuffer,
  isHDv3WalletByType,
  cachedXpubByType,
  cachedHdV3KeyStoreByType,
  accountAddresses,
  isPublicAccount,
  isHardwareAccount,
  balance,
  getHdWalletBySeed,
};
