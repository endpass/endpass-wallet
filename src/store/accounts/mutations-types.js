export const CHANGE_INIT_STATUS = 'CHANGE_INIT_STATUS';
export const SET_ADDRESS = 'SET_ADDRESS';
export const ADD_WALLET = 'ADD_WALLET';
export const SET_HD_KEY = 'SET_HD_KEY';
export const SET_BALANCE = 'SET_BALANCE';
export const SET_HD_CACHE_BY_TYPE = 'SET_HD_CACHE_BY_TYPE';

export default {
  CHANGE_INIT_STATUS,
  SET_ADDRESS,
  ADD_WALLET,
  SET_HD_KEY,
  SET_BALANCE,
  SET_HD_CACHE_BY_TYPE,
};
