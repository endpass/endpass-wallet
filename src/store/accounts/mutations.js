import { Wallet, Address } from '@/class';
import {
  ADD_ADDRESS,
  SET_ADDRESS,
  ADD_WALLET,
  SET_WALLET,
  SET_HD_KEY,
  SET_BALANCE,
} from './mutations-types';

const setAddress = (state, newAddress) => {
  let address = newAddress;

  if (!(address instanceof Address)) {
    address = new Address({ address });
  }

  state.address = address;
};

const setWallet = (state, wallet) => {
  state.wallet = wallet;
};

const addWallet = (state, walletV3) => {
  const wallet = new Wallet(walletV3);
  const { address } = walletV3;
  state.wallets = {
    ...state.wallets,
    [address]: wallet,
  };
};

// Adds an empty Address as wallet to use public wallet functionality
const addAddress = (state, { address, info }) => {
  state.wallets = {
    ...state.wallets,
    [address]: new Address({ address, info }),
  };
};

// Saves the encrypted HD wallet key in V3 keystore format
// Formerly addHdWallet
const setHdKey = (state, key) => {
  state.hdKey = key;
};

const setBalance = (state, balance) => {
  state.balance = balance;
};

export default {
  [ADD_ADDRESS]: addAddress,
  [SET_ADDRESS]: setAddress,
  [ADD_WALLET]: addWallet,
  [SET_WALLET]: setWallet,
  [SET_HD_KEY]: setHdKey,
  [SET_BALANCE]: setBalance,
};
