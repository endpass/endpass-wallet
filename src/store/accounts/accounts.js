import { userService } from '@/services';
import storage from '@/services/storage';
import web3 from 'web3';
import Bip39 from 'bip39';
import HDKey from 'ethereumjs-wallet/hdkey';
import { hdKeyMnemonic, kdfParams } from '@/config';
import EthWallet from 'ethereumjs-wallet';
import { SAVE_TOKENS } from '@/store/tokens/mutations-types';
import { Wallet, Address } from '@/class';
import { BigNumber } from 'bignumber.js';
import keystore from '@/utils/keystore';

export default {
  namespaced: true,
  state: {
    // TODO move to user store?
    email: null,
    // The encrypted xprv key for the user's hd wallet as a V3 keystore
    // object. All accounts are direct child accounts.
    hdKey: null,
    // The user's wallets, keyed by address
    wallets: {},
    // The currently selected wallet
    wallet: null,
    address: null,
    balance: null,
    price: null,
    // prettier-ignore
    availableCurrencies: ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
    settings: {
      fiatCurrency: 'USD',
    },
    otpSettings: {
      secret: null,
      status: null,
    },
  },
  getters: {
    isPublicAccount(state) {
      return (
        state.address instanceof Address && !(state.wallet instanceof Wallet)
      );
    },
    // TODO move to user store?
    isLoggedIn(state) {
      return !!state.email;
    },
    balance(state, getters, rootState, rootGetters) {
      if (state.balance === null) return null;

      const pendingBalance = rootGetters['transactions/pendingBalance'];
      const balanceWei = BigNumber(state.balance || '0')
        .minus(pendingBalance)
        .toFixed();

      return web3.utils.fromWei(balanceWei);
    },
    // Returns a decrypted HD Wallet
    hdWallet: state => password => {
      if (!state.hdKey) {
        return null;
      }
      return keystore.decryptHDWallet(password, state.hdKey);
    },
  },
  mutations: {
    setAddress(state, addressString) {
      state.address = new Address(addressString);
    },
    setWallet(state, wallet) {
      state.wallet = wallet;
    },
    addWallet(state, walletV3) {
      const wallet = new Wallet(walletV3);
      const { address } = walletV3;
      state.wallets = {
        ...state.wallets,
        [address]: wallet,
      };
    },
    // Saves the encrypted HD wallet key in V3 keystore format
    // Formerly addHdWallet
    setHdKey(state, key) {
      state.hdKey = key;
    },
    setBalance(state, balance) {
      state.balance = balance;
    },
    setPrice(state, price) {
      state.price = price;
    },
    setSettings(state, settings) {
      state.settings = JSON.parse(JSON.stringify(settings));
    },
    setEmail(state, email) {
      state.email = email;
    },
    setOtpSettings(state, otpSettings) {
      state.otpSettings = otpSettings;
    },
  },
  actions: {
    selectWallet({ commit, state, dispatch }, address) {
      commit('setWallet', state.wallets[address]);
      commit('setAddress', address);
      return dispatch(
        'tokens/subscribeOnTokensBalancesUpdates',
        {},
        { root: true },
      );
    },
    addWallet({ commit, dispatch, state }, json) {
      json.address = web3.utils.toChecksumAddress(json.address);
      commit('addWallet', json);

      return userService
        .setAccount(json.address, json)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    addWalletAndSelect({ dispatch }, json) {
      return dispatch('addWallet', json)
        .then(() =>
          dispatch('selectWallet', web3.utils.toChecksumAddress(json.address)),
        )
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    // Import wallet from json V3 keystore
    async addWalletWithV3({ commit, dispatch }, { json, password }) {
      try {
        const wallet = new Wallet(json);
        let privateKey = await wallet.getPrivateKey(password);
        return dispatch('addWalletWithPrivateKey', privateKey);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    async addWalletWithPrivateKey(
      { commit, dispatch },
      { privateKey, password },
    ) {
      try {
        const wallet = EthWallet.fromPrivateKey(Buffer.from(privateKey, 'hex'));
        const json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    generateWallet({ commit, dispatch, state, getters }, password) {
      if (!state.hdKey) {
        return;
      }
      try {
        let hdWallet = getters.hdWallet(password);
        let i = Object.keys(state.wallets).length;
        let wallet = hdWallet.deriveChild(i).getWallet();
        let json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    // Saves HD wallet's extended keys on the server
    saveHdWallet({ commit, dispatch, state }, json) {
      return userService.setAccount(json.address, json);
    },
    addHdWallet({ commit, dispatch }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);
      // Encrypt extended private key
      const json = keystore.encryptHDWallet(password, hdWallet);
      commit('setHdKey', json);

      // Save HD keys and generate the first child wallet
      return dispatch('saveHdWallet', json)
        .then(() => dispatch('generateWallet', password))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async addMultiHdWallet({ commit, dispatch, rootState }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);

      /* eslint-disable no-await-in-loop */
      for (let index = 0; index < 5; index++) {
        const wallet = hdWallet.deriveChild(index).getWallet();
        const walletV3 = wallet.toV3(Buffer.from(password), kdfParams);
        const { address } = walletV3;

        if (index !== 0) {
          dispatch('addWallet', walletV3);
        } else {
          dispatch('addWalletAndSelect', walletV3);
        }

        try {
          const balance = await rootState.web3.web3.eth.getBalance(address);

          if (balance === '0') {
            break;
          }
        } catch (e) {
          break;
        }
      }
      /* eslint-enable no-await-in-loop */
    },
    updateBalance({ commit, dispatch, state, rootState }) {
      if (state.address) {
        const address = state.address.getChecksumAddressString();

        return rootState.web3.web3.eth
          .getBalance(address)
          .then(balance => commit('setBalance', balance))
          .catch(e => dispatch('errors/emitError', e, { root: true }));
      }
    },
    updateSettings({ commit, dispatch }, settings) {
      commit('setSettings', settings);
      return storage
        .write('settings', settings)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    validatePassword({ state }, password) {
      return state.wallet.validatePassword(password);
    },
    login({ commit, dispatch }, email) {
      return userService.login(email);
    },
    logout({ commit, dispatch }) {
      commit('setEmail', null);
      return Promise.all([storage.clear(), userService.logout()])
        .then(() => window.location.reload())
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    loginViaOTP({}, { code, email }) {
      return userService.loginViaOTP(code, email);
    },
    getOtpSettings({ commit, dispatch }) {
      return userService
        .getOtpSettings()
        .then(otpSettings => commit('setOtpSettings', otpSettings))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    setOtpSettings({ commit, dispatch }, { secret, code }) {
      return userService
        .setOtpSettings(secret, code)
        .then(() => commit('setOtpSettings', { status: 'enabled' }))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    deleteOtpSettings({ commit, dispatch }, { code }) {
      return userService
        .deleteOtpSettings(code)
        .then(() => {
          commit('setOtpSettings', {});
          dispatch('getOtpSettings');
        })
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async init({ commit, dispatch }) {
      try {
        let [settings, email] = await Promise.all([
          userService.getSettings(),
          storage.read('email'),
        ]);

        commit(`tokens/${SAVE_TOKENS}`, settings.tokens || {}, { root: true });
        commit('setEmail', email);

        if (settings) {
          commit('setSettings', settings);
        }

        if (!email) {
          storage.disableRemote();
          return null;
        }

        // Fetch and save HD wallet
        let hdKey = await userService.getHDKey();
        if (hdKey) {
          commit('setHdKey', hdKey);
        }

        // Fetch and save regular accounts
        let accounts = await userService.getV3Accounts();
        if (accounts && accounts.length) {
          accounts.forEach(wallet => commit('addWallet', wallet));
          await dispatch('selectWallet', accounts[0].address);
        }
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
  },
};
