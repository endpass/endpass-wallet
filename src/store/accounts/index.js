import actions from './actions';
import getters from './getters';
import mutations from './mutations';

/**
 * @namespace
 * @property {Boolean} isInited Status of initializing accounts module
 * @property {Object} hdKey The encrypted xprv key for the user's hd wallet as a V3 keystore object.
 *  All accounts are direct child accounts.
 * @property {Object} hadrwareXpub The user's hardware xpub, keyed by wallet type
 * @property {Object} wallets The user's wallets, mapped by address
 * @property {String} address Current wallet address
 * @property {String} balance Currect account balance
 */
const state = {
  isInited: false,
  hardwareXpub: {},
  wallets: {},
  hdKey: null,
  address: null,
  balance: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
