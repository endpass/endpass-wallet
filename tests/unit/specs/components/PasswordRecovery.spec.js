import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import Notifications from 'vue-notification';
import UIComponents from '@endpass/ui';

import PasswordRecovery from '@/components/PasswordRecovery';
import { mnemonic } from 'fixtures/accounts';

describe('PasswordRecovery', () => {
  let accountsActions;
  let storeOptions;
  let wrapper;

  beforeEach(() => {
    accountsActions = {
      recoverWalletsPassword: jest.fn().mockResolvedValue(true),
    };
    storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions: accountsActions,
        },
      },
    };
    const localVue = createLocalVue();

    localVue.use(Vuex);
    localVue.use(Notifications);
    localVue.use(VeeValidate);
    localVue.use(UIComponents);

    const store = new Vuex.Store(storeOptions);

    wrapper = shallowMount(PasswordRecovery, {
      store,
      localVue,
      sync: false,
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('PasswordRecovery');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render the initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('seed phrase field', () => {
      it('should disable field', async () => {
        expect.assertions(1);

        wrapper.setData({
          isLoading: true,
        });

        await wrapper.vm.$nextTick();

        expect(
          wrapper.find('v-input-stub[data-test=input-seed-phrase]').attributes()
            .disabled,
        ).toBeTruthy();
      });

      it('should enable field', async () => {
        expect.assertions(1);

        wrapper.setData({
          isLoading: false,
        });

        await wrapper.vm.$nextTick();

        expect(
          wrapper.find('v-input-stub[data-test=input-seed-phrase]').attributes()
            .disabled,
        ).toBeUndefined();
      });
    });

    describe('password field', () => {
      it('should disable field', async () => {
        expect.assertions(1);

        wrapper.setData({
          isLoading: true,
        });

        await wrapper.vm.$nextTick();

        expect(
          wrapper.find('v-password-stub[data-test=input-password]').attributes()
            .disabled,
        ).toBeTruthy();
      });

      it('should enable field', async () => {
        expect.assertions(1);

        wrapper.setData({
          isLoading: false,
        });

        await wrapper.vm.$nextTick();

        expect(
          wrapper.find('v-password-stub[data-test=input-password]').attributes()
            .disabled,
        ).toBeUndefined();
      });
    });

    describe('button', () => {
      it('should show loader', async () => {
        expect.assertions(1);

        wrapper.setData({
          isLoading: true,
        });

        await wrapper.vm.$nextTick();

        expect(wrapper.find('v-button-stub').attributes().loading).toBeTruthy();
      });

      it('should not show loader', async () => {
        expect.assertions(1);

        wrapper.setData({
          isLoading: false,
        });

        await wrapper.vm.$nextTick();

        expect(
          wrapper.find('v-button-stub').attributes().loading,
        ).toBeUndefined();
      });
    });

    describe('confirm modal', () => {
      it('should toggle confirm modal', async () => {
        wrapper.vm.toggleConfirmModal();
        await wrapper.vm.$nextTick();
        expect(wrapper.find('confirm-modal-stub').html()).toMatchSnapshot();

        wrapper.vm.toggleConfirmModal();
        await wrapper.vm.$nextTick();
        expect(wrapper.find('confirm-modal-stub').exists()).toBeFalsy();
      });
    });
  });

  describe('behavior', () => {
    const password = 'password';
    const seedPhrase = 'seed phrase';

    describe('form', () => {
      it('should open confirm window', async () => {
        expect.assertions(1);

        wrapper.setData({
          seedPhrase,
          password,
        });

        wrapper.find('v-form-stub').vm.$emit('submit');

        await wrapper.vm.$nextTick();

        expect(wrapper.find('confirm-modal-stub').exists()).toBeTruthy();
      });
    });

    describe('confirm modal', () => {
      beforeEach(async () => {
        jest.spyOn(wrapper.vm, '$notify');

        wrapper.setData({
          seedPhrase,
          password,
          isConfirmModal: true,
        });

        await wrapper.vm.$nextTick();
      });

      it('should recover wallets password', async () => {
        expect.assertions(7);

        wrapper.find('confirm-modal-stub').vm.$emit('confirm');

        expect(wrapper.vm.isLoading).toBe(true);

        await wrapper.vm.$nextTick();

        expect(accountsActions.recoverWalletsPassword).toHaveBeenCalledTimes(1);
        expect(accountsActions.recoverWalletsPassword).toHaveBeenCalledWith(
          expect.any(Object),
          { password, seedPhrase },
          undefined,
        );
        expect(wrapper.vm.$notify).toHaveBeenCalledTimes(0);
        expect(wrapper.vm.isLoading).toBe(false);
        expect(wrapper.vm.seedPhrase).toBe(null);
        expect(wrapper.vm.password).toBe(null);
      });

      it('should handle exceptions', async () => {
        expect.assertions(8);

        accountsActions.recoverWalletsPassword.mockRejectedValueOnce();

        wrapper.find('confirm-modal-stub').vm.$emit('confirm');

        expect(wrapper.vm.isLoading).toBe(true);

        await wrapper.vm.$nextTick();

        expect(accountsActions.recoverWalletsPassword).toHaveBeenCalledTimes(1);
        expect(accountsActions.recoverWalletsPassword).toHaveBeenCalledWith(
          expect.any(Object),
          { password, seedPhrase },
          undefined,
        );
        expect(wrapper.vm.$notify).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$notify).toHaveBeenCalledWith({
          title: 'Error recovering wallet password',
          text:
            'An error occurred while recovering wallet password. Please try again.',
          type: 'is-danger',
        });
        expect(wrapper.vm.isLoading).toBe(false);
        expect(wrapper.vm.seedPhrase).toBe(null);
        expect(wrapper.vm.password).toBe(null);
      });

      it('should close confirm modal', async () => {
        expect.assertions(1);

        wrapper.find('confirm-modal-stub').vm.$emit('close');
        await wrapper.vm.$nextTick();

        expect(wrapper.find('confirm-modal-stub').exists()).toBeFalsy();
      });
    });
  });
});
