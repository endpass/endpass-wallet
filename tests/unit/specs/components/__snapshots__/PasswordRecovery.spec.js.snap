// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordRecovery render confirm modal should toggle confirm modal 1`] = `
<confirm-modal-stub>
  <p class="subtitle">
    All keystore wallets will be deleted. They can be restored from a
    private key or seed phrase on the import page.
  </p>
</confirm-modal-stub>
`;

exports[`PasswordRecovery render should render the initial state of the component 1`] = `
<div>
  <v-form-stub
    data-test="password-recovery-form"
  >
    <label
      class="label"
    >
      Password Recovery
    </label>
     
    <v-input-stub
      aria-invalid="false"
      aria-required="true"
      classname=""
      data-test="input-seed-phrase"
      name="seedPhrase"
      placeholder="Seed phrase"
    />
     
    <v-password-stub
      aria-invalid="false"
      aria-required="true"
      data-test="input-password"
      name="password"
      placeholder="New password"
    />
     
    <v-button-stub
      classname="is-primary is-medium"
      data-test="recover-button"
      disabled="true"
    >
      
      Recover
    
    </v-button-stub>
  </v-form-stub>
   
  <!---->
</div>
`;
